File: backend/internal/fracindex/fracindex.go
package fracindex

import "fmt"

const (
	minByte = byte(0x21) // '!'
	maxByte = byte(0x7E) // '~'
)

func GenerateKeyBetween(a, b string) (string, error) {
	if a != "" && b != "" && a >= b {
		return "", fmt.Errorf("invalid bounds: %q >= %q", a, b)
	}
	switch {
	case a == "" && b == "":
		return string(midByte(minByte, maxByte)), nil
	case a == "":
		return before(b), nil
	case b == "":
		return after(a), nil
	default:
		return between(a, b), nil
	}
}

func between(a, b string) string {
	n := min(len(a), len(b))
	for i := 0; i < n; i++ {
		if a[i] != b[i] {
			if b[i]-a[i] > 1 {
				return a[:i] + string(midByte(a[i], b[i]))
			}
			return a[:i+1] + after(a[i+1:])
		}
	}
	return a + before(b[len(a):])
}

func before(b string) string {
	for i := 0; i < len(b); i++ {
		if b[i] > minByte {
			mid := midByte(minByte, b[i])
			if mid > minByte {
				return b[:i] + string(mid)
			}
			return b[:i] + string(minByte) + string(midByte(minByte, maxByte))
		}
	}
	return b + string(midByte(minByte, maxByte))
}

func after(a string) string {
	for i := len(a) - 1; i >= 0; i-- {
		if a[i] < maxByte {
			return a[:i] + string(midByte(a[i], maxByte))
		}
	}
	return a + string(midByte(minByte, maxByte))
}

func midByte(lo, hi byte) byte {
	return lo + (hi-lo)/2
}

File: frontend/src/lib/fracindex.ts
const MIN_CHAR_CODE = 0x21; // '!'
const MAX_CHAR_CODE = 0x7e; // '~'

export function generateKeyBetween(a: string, b: string): string {
  if (a !== '' && b !== '' && a >= b) {
    throw new Error(`invalid bounds: "${a}" >= "${b}"`);
  }
  if (a === '' && b === '') return midChar(MIN_CHAR_CODE, MAX_CHAR_CODE);
  if (a === '') return before(b);
  if (b === '') return after(a);
  return between(a, b);
}

function between(a: string, b: string): string {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const ca = a.charCodeAt(i);
    const cb = b.charCodeAt(i);
    if (ca !== cb) {
      if (cb - ca > 1) {
        return a.slice(0, i) + midChar(ca, cb);
      }
      return a.slice(0, i + 1) + after(a.slice(i + 1));
    }
  }
  return a + before(b.slice(a.length));
}

function before(b: string): string {
  for (let i = 0; i < b.length; i++) {
    const c = b.charCodeAt(i);
    if (c > MIN_CHAR_CODE) {
      const mid = midCode(MIN_CHAR_CODE, c);
      if (mid > MIN_CHAR_CODE) {
        return b.slice(0, i) + String.fromCharCode(mid);
      }
      return b.slice(0, i) + String.fromCharCode(MIN_CHAR_CODE) + midChar(MIN_CHAR_CODE, MAX_CHAR_CODE);
    }
  }
  return b + midChar(MIN_CHAR_CODE, MAX_CHAR_CODE);
}

function after(a: string): string {
  for (let i = a.length - 1; i >= 0; i--) {
    const c = a.charCodeAt(i);
    if (c < MAX_CHAR_CODE) {
      return a.slice(0, i) + midChar(c, MAX_CHAR_CODE);
    }
  }
  return a + midChar(MIN_CHAR_CODE, MAX_CHAR_CODE);
}

function midCode(lo: number, hi: number): number {
  return lo + Math.floor((hi - lo) / 2);
}

function midChar(lo: number, hi: number): string {
  return String.fromCharCode(midCode(lo, hi));
}
